### PR TITLE
LibFileSystem+Everywhere: Add `LibFileSystem`

### PR DESCRIPTION
**Pull request by cammo1123**
Port existing functions from `DeprecatedFile` to a new library `LibFileSystem`. This also changes all trivial code to use the new functions, and removes all now unused functions from `DeprecatedFile`

### CODE REVIEW

**Comment by timschumi** [comment]
It would be great if we could keep `Core::File` exclusive to the stream-based file implementation instead of also putting a bunch of utility functions in there (for example, `is_directory` doesn't really make sense here anyways, you can't read from it, so you'd never construct a `Core::File` from it in the first place).

I briefly talked about this to other people while drafting up the `Core::Stream::File` -> `Core::File` migration process, the most reasonable suggestion I have heard of so far is to move those utility functions in a new class called `Core::Filesystem` (or `Core::FileSystem`?) and to drop the variations of the functions that don't take either an `fd` or a `filename`.

**Comment by kleinesfilmroellchen** [comment]
> It would be great if we could keep `Core::File` exclusive to the stream-based file implementation instead of also putting a bunch of utility functions in there (for example, `is_directory` doesn't really make sense here anyways, you can't read from it, so you'd never construct a `Core::File` from it in the first place).
> 
> I briefly talked about this to other people while drafting up the `Core::Stream::File` -> `Core::File` migration process, the most reasonable suggestion I have heard of so far is to move those utility functions in a new class called `Core::Filesystem` (or `Core::FileSystem`?) and to drop the variations of the functions that don't take either an `fd` or a `filename`.

Exactly my thoughts, the name `Filesystem` sounds good.

**Comment by cammo1123** [comment]
Absolutely! I've marked it as a draft for now while I figure out how big the scope should be, I'm thinking all the `copy_file_or_directory` stuff should also go into it too! Honestly all the static functions inside `DeprecatedFile` seem like they'd fit perfectly aswell

**Comment by LucasChollet** [comment]
If there is only `static` methods in it, `FileSystem` should be a namespace.

**Comment by cammo1123** [comment]
Done, Done and Done! It only ended up needing to be a namespace in the end. Much happier with how looks/feels now!

**Comment by timschumi** [comment]
Looking better! That said, there might be a point in gradually migrating everything while modernizing the functions a bit. Ideally, that would include things like changing leftover `DeprecatedString const&` to `StringView`, and (if deemed reasonable) maybe some more error-awareness and/or porting work from `DeprecatedString` to the new `String`.

**Comment by cammo1123** [comment]
> Looking better! That said, there might be a point in gradually migrating everything while modernizing the functions a bit. Ideally, that would include things like changing leftover `DeprecatedString const&` to `StringView`, and (if deemed reasonable) maybe some more error-awareness and/or porting work from `DeprecatedString` to the new `String`.

I gave it a once over and added a commit modernizing it. More to be done, but it feels like a significant first step.

File: Userland/Libraries/LibCore/FileSystem.cpp [hunk 1]
```diff
@@ -231,15 +240,16 @@ ErrorOr<void, CopyError> copy_file(DeprecatedString const& dst_path, DeprecatedS
...
             return CopyError { errno, false };
     }
     return {};
 }
 
-ErrorOr<void, CopyError> copy_directory(DeprecatedString const& dst_path, DeprecatedString const& src_path, struct stat const& src_stat, LinkMode link, PreserveMode preserve_mode)
+ErrorOr<void, CopyError> copy_directory(StringView dst_path, StringView src_path, struct stat const& src_stat, LinkMode link, PreserveMode preserve_mode)
 {
-    if (mkdir(dst_path.characters(), 0755) < 0)
+    if (System::mkdir(dst_path, 0755).is_error())
```
Comment (timschumi) [inline Userland/Libraries/LibCore/FileSystem.cpp:252]: Other places use `Core::System` instead of just `System`, so we should probably keep that as consistent as possible.

File: Userland/Libraries/LibCore/FileSystem.cpp [hunk 2]
```diff
@@ -231,15 +240,16 @@ ErrorOr<void, CopyError> copy_file(DeprecatedString const& dst_path, DeprecatedS
             src_stat.st_mtim,
 #endif
         };
-        if (utimensat(AT_FDCWD, dst_path.characters(), times, 0) < 0)
+        DeprecatedString dst_path_string = dst_path;
+        if (utimensat(AT_FDCWD, dst_path_string.characters(), times, 0) < 0)
```
Comment (timschumi) [inline Userland/Libraries/LibCore/FileSystem.cpp:244]: `utimensat` is one of the chonkier LibC syscall wrappers, but we could technically add a `Core::System::utimensat` that takes a `StringView` and sends the string to the kernel without copying.

File: Userland/Libraries/LibCore/FileSystem.cpp [hunk 3]
```diff
@@ -267,11 +277,11 @@ ErrorOr<void, CopyError> copy_directory(DeprecatedString const& dst_path, Deprec
     auto my_umask = umask(0);
     umask(my_umask);
 
-    if (chmod(dst_path.characters(), src_stat.st_mode & ~my_umask) < 0)
+    if (System::chmod(dst_path, src_stat.st_mode & ~my_umask).is_error())
         return CopyError { errno, false };
```
Comment (timschumi) [inline Userland/Libraries/LibCore/FileSystem.cpp]: I'm not sure whether we can/should trust `errno` when using the `Core::System` wrappers, so we should probably store the returned `ErrorOr` and use `.error().errno()` (or however we named that function) instead.

File: Userland/Libraries/LibCore/FileSystem.cpp (hunk 1 above)
Comment (AtkinsSJ) [reply Userland/Libraries/LibCore/FileSystem.cpp:252]: IMO, it's fine not to when we're already in the namespace, but we're a bit inconsistent throughout the codebase of whether we do or don't.

Comment (cammo1123) [reply Userland/Libraries/LibCore/FileSystem.cpp]: I've made it consistently use `System::` as that seemed reasonable. If it needs to be the `Core::System` I'm happy to do that to!

Comment (timschumi) [reply Userland/Libraries/LibCore/FileSystem.cpp]: I meant to express that we should pick one of both ("other places" being other places in the same file) and use that consistently, so this is fine. :^)

File: Userland/Libraries/LibCore/System.cpp [hunk 4]
```diff
@@ -1055,6 +1055,28 @@ ErrorOr<void> utime(StringView path, Optional<struct utimbuf> maybe_buf)
 #endif
 }
 
+ErrorOr<void> utimensat(int fd, StringView path, struct timespec const times[2], int flag)
+{
+    if (path.is_null())
+        return Error::from_errno(EFAULT);
+
+#ifdef AK_OS_SERENITY
+    Syscall::SC_utimensat_params params {
```
Comment (timschumi) [inline Userland/Libraries/LibCore/System.cpp:1064]: Do we not need any of the checks that are listed here, at least for the SerenityOS case?

https://github.com/SerenityOS/serenity/blob/82219d0879aee5d47abccfeab6b7ca5894e5e522/Userland/Libraries/LibC/fcntl.cpp#L137-L162

File: Userland/Libraries/LibCore/FileSystem.cpp [hunk 5]
```diff
@@ -285,14 +294,15 @@ ErrorOr<void, CopyError> copy_directory(DeprecatedString const& dst_path, Deprec
             src_stat.st_mtim,
 #endif
         };
-        if (utimensat(AT_FDCWD, dst_path.characters(), times, 0) < 0)
-            return CopyError { errno, false };
+        DeprecatedString dst_path_string = dst_path;
```
Comment (timschumi) [inline Userland/Libraries/LibCore/FileSystem.cpp:297]: This seems unused now.

File: Userland/Applications/Run/RunWindow.cpp [hunk 6]
```diff
@@ -142,9 +142,9 @@ bool RunWindow::run_via_launch(DeprecatedString const& run_input)
     auto url = URL::create_with_url_or_path(run_input);
 
     if (url.scheme() == "file") {
-        auto real_path = Core::DeprecatedFile::real_path_for(url.path());
+        auto real_path = Core::FileSystem::real_path(url.path());
         if (real_path.is_null()) {
-            // errno *should* be preserved from Core::DeprecatedFile::real_path_for().
+            // errno *should* be preserved from Core::FileSystem::real_path().
```
Comment (timschumi) [inline Userland/Applications/Run/RunWindow.cpp:147]: It's slightly out of scope for this commit, but while we are here anyways, we should probably make this a `// FIXME: Make real_path use ErrorOr<> and use that to get the error code`.

While we might be able to do that inside the implementation itself, we shouldn't be assuming that our more modern wrappers set or retain errno correctly to the outside.

**Review by timschumi — CHANGES_REQUESTED** [review]
Three remaining very small things.

File: Userland/Libraries/LibCore/FileSystem.h [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+
+#include <AK/DeprecatedString.h>
+#include <AK/Error.h>
+#include <AK/StringView.h>
+#include <LibCore/File.h>
+#include <sys/stat.h>
+
+namespace Core::FileSystem {
+
+#define DEFAULT_PATH "/usr/local/sbin:/usr/local/bin:/usr/bin:/bin"
```
Comment (pascalpuffke) [inline Userland/Libraries/LibCore/FileSystem.h:17]: If it isn't entirely out of scope, can we not have `DEFAULT_PATH` as preprocessor defines?

`static constexpr StringView default_path = "/usr/local/sbin:/usr/local/bin:/usr/bin:/bin"sv` will work fine for all `DEFAULT_PATH_SV` uses, and the few occurences in LibC/unistd.cpp and TelnetServer/main.cpp can use `.characters()`

File: Userland/Libraries/LibC/unistd.cpp [hunk 8]
```diff
@@ -9,7 +9,7 @@
 #include <AK/ScopedValueRollback.h>
 #include <AK/Vector.h>
 #include <Kernel/API/Unveil.h>
-#include <LibCore/DeprecatedFile.h>
+#include <LibCore/FileSystem.h>
```
Comment (kleinesfilmroellchen) [inline Userland/Libraries/LibC/unistd.cpp:12]: This include seems to be unused, let's remove it.

File: Userland/Libraries/LibCore/FileSystem.cpp [hunk 9]
```diff
@@ -0,0 +1,449 @@
+/*
+ * Copyright (c) 2023, Cameron Youell <cameronyouell@gmail.com>
+ * Copyright (c) 2018-2022, Andreas Kling <kling@serenityos.org>
```
Comment (kleinesfilmroellchen) [inline Userland/Libraries/LibCore/FileSystem.cpp:3]: This order is wrong.

File: Userland/Libraries/LibCore/FileSystem.cpp [hunk 10]
```diff
@@ -0,0 +1,501 @@
...
+#endif
+
+// On Linux distros that use glibc `basename` is defined as a macro that expands to `__xpg_basename`, so we undefine it
+#if defined(AK_OS_LINUX) && defined(basename)
+#    undef basename
+#endif
+
+namespace Core::FileSystem {
+
+DeprecatedString current_working_directory()
```
[the remark this answers is not in the document]
Comment (kleinesfilmroellchen) [reply Userland/Libraries/LibCore/FileSystem.cpp:24]: As a more precise request: Please port this function to ErrorOr<String> and use Core::System.

File: Userland/Libraries/LibCore/FileSystem.cpp [hunk 11]
```diff
@@ -0,0 +1,449 @@
...
+        return {};
+    }
+
+    auto cwd_as_string = DeprecatedString(cwd);
+    free(cwd);
+
+    return cwd_as_string;
+}
+
+DeprecatedString absolute_path(StringView path)
```
Comment (kleinesfilmroellchen) [inline Userland/Libraries/LibCore/FileSystem.cpp:38]: Error port

File: Userland/Libraries/LibCore/FileSystem.cpp [hunk 12]
```diff
@@ -0,0 +1,449 @@
...
+
+bool is_link(int fd)
+{
+    struct stat st;
+    if (fstat(fd, &st) < 0)
+        return false;
+    return S_ISLNK(st.st_mode);
+}
+
+static DeprecatedString get_duplicate_name(DeprecatedString const& path, int duplicate_count)
```
Comment (kleinesfilmroellchen) [inline Userland/Libraries/LibCore/FileSystem.cpp:156]: ErrorOr<String> and stuff :^)

File: Userland/Libraries/LibCore/FileSystem.cpp [hunk 13]
```diff
@@ -0,0 +1,449 @@
...
+        title = title.substring_view(0, title.length() - prev_duplicate_tag.length());
+    }
+    duplicated_name.appendff("{} ({})", title, duplicate_count);
+    if (!lexical_path.extension().is_empty()) {
+        duplicated_name.appendff(".{}", lexical_path.extension());
+    }
+    return duplicated_name.to_deprecated_string();
+}
+
+ErrorOr<void, CopyError> copy_file(StringView dst_path, StringView src_path, struct stat const& src_stat, NonnullOwnPtr<File>& source, PreserveMode preserve_mode)
```
Comment (kleinesfilmroellchen) [inline Userland/Libraries/LibCore/FileSystem.cpp:181]: Taking a reference to a NonnullOwnPtr is not a good idea, I think you can take a regular reference instead which NonnullOwnPtr can provide without problems.

File: Userland/Libraries/LibCore/FileSystem.cpp [hunk 14]
```diff
@@ -0,0 +1,449 @@
...
+    bool is_directory_sticky = directory_stat.st_mode & S_ISVTX;
+    if (!is_directory_sticky)
+        return true;
+
+    // Directory is sticky, only the file owner, directory owner, and root can modify (rename, remove) it.
+    auto user_id = geteuid();
+    return user_id == 0 || directory_stat.st_uid == user_id || stat_or_empty(path).st_uid == user_id;
+}
+
+ErrorOr<DeprecatedString> read_link(StringView link_path)
```
Comment (kleinesfilmroellchen) [inline Userland/Libraries/LibCore/FileSystem.cpp:395]: You can easily construct a String here.

**Review by kleinesfilmroellchen — CHANGES_REQUESTED** [review]
I think I looked at everything now. If you move functions to FileSystem, I believe that they should immediately be in a fully modern state, anything else is not worth it. Since most functions already return errors, it should be an easy task of returning a different string type.

**Review by kleinesfilmroellchen — CHANGES_REQUESTED** [review]
Much better! Just a few more things.

File: Userland/Libraries/LibCore/FileSystem.h [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+
+enum class PreserveMode {
+    Nothing = 0,
+    Permissions = (1 << 0),
+    Ownership = (1 << 1),
+    Timestamps = (1 << 2),
+};
+AK_ENUM_BITWISE_OPERATORS(PreserveMode);
+
+ErrorOr<void> copy_file(StringView dst_path, StringView src_path, struct stat const& src_stat, File* source, PreserveMode = PreserveMode::Nothing);
```
[the remark this answers is not in the document]
Comment (cammo1123) [reply Userland/Libraries/LibCore/FileSystem.h]: Correct!

File: Userland/Libraries/LibCore/System.cpp [hunk 16]
```diff
@@ -1055,6 +1055,52 @@ ErrorOr<void> utime(StringView path, Optional<struct utimbuf> maybe_buf)
...
+        .dirfd = fd,
+        .path = { path.characters_without_null_termination(), path.length() },
+        .times = times,
+        .flag = flag,
+    };
+    int rc = syscall(SC_utimensat, &params);
+    HANDLE_SYSCALL_RETURN_VALUE("utimensat", rc, {});
+#else
+    DeprecatedString path_string = path;
```
[the remark this answers is not in the document]
Comment (cammo1123) [reply Userland/Libraries/LibCore/System.cpp]: I'm really not sure how I should go about this change

File: Userland/Libraries/LibCore/File.h [hunk 17]
```diff
@@ -68,6 +68,12 @@ class File final : public SeekableStream {
     virtual ErrorOr<size_t> seek(i64 offset, SeekMode) override;
     virtual ErrorOr<void> truncate(size_t length) override;
 
+    bool is_directory() const;
+    bool is_device() const;
+    bool is_block_device() const;
+    bool is_char_device() const;
+    bool is_link() const;
+
```
Comment (timschumi) [inline Userland/Libraries/LibCore/File.h:76]: I'm not sure how I feel about having those functions here, they would presumably be much more universal if they were in `Core::FileSystem` and taking an `fd` argument. `Core::File` already has an `fd` function that returns the currently open file descriptor, so those can be easily combined without needing to put them into the `Core::File` class itself.

File: Userland/Applications/PartitionEditor/PartitionModel.h [hunk 18]
```diff
@@ -6,6 +6,7 @@
 
 #pragma once
 
+#include <LibCore/DeprecatedFile.h>
```
Comment (timschumi) [inline Userland/Applications/PartitionEditor/PartitionModel.h:9]: This doesn't seem to be used in this header at all.

File: Userland/Libraries/LibCore/FileSystem.cpp [hunk 19]
```diff
@@ -1,5 +1,6 @@
 /*
  * Copyright (c) 2023, Cameron Youell <cameronyouell@gmail.com>
+ * Copyright (c) 2020-2023, the SerenityOS developers.
```
Comment (timschumi) [inline Userland/Libraries/LibCore/FileSystem.cpp]: This should probably be the first entry, due to it starting earlier.

File: Userland/Libraries/LibCore/FileSystem.h [hunk 20]
```diff
@@ -1,5 +1,6 @@
 /*
  * Copyright (c) 2023, Cameron Youell <cameronyouell@gmail.com>
+ * Copyright (c) 2020-2023, the SerenityOS developers.
```
Comment (timschumi) [inline Userland/Libraries/LibCore/FileSystem.h]: Same here.

File: Userland/Libraries/LibCore/FileSystem.h [hunk 21]
```diff
@@ -17,6 +18,30 @@ namespace Core::FileSystem {
 #define DEFAULT_PATH "/usr/local/sbin:/usr/local/bin:/usr/bin:/bin"
 #define DEFAULT_PATH_SV "/usr/local/sbin:/usr/local/bin:/usr/bin:/bin"sv
 
+class TempFile {
+    friend ErrorOr<NonnullOwnPtr<TempFile>> create_temp_directory();
+    friend ErrorOr<NonnullOwnPtr<TempFile>> create_temp_file();
+
+public:
+    ~TempFile();
+    String const& path() const { return m_path; }
```
Comment (timschumi) [inline Userland/Libraries/LibCore/FileSystem.h:27]: This should maybe return a StringView instead.

**Review by timschumi — CHANGES_REQUESTED** [review]
One last round of nitpicks from me, after that it should be in pretty good shape. :^)

**Comment by cammo1123** [comment]
> One last round of nitpicks from me, after that it should be in pretty good shape. :^)

All were very simple changes to make. Done!

**Review by timschumi — APPROVED** [review]
LGTM

**Review by cbyrneee — COMMENTED** [review]
I took a general look at the `Core::Filesystem` functions, I didn't look at the usage of it in our existing applications in much depth as it seemed to be pretty trivial cases that were changed (as stated in your PR).

This is a very nice change to see, and the introduction of more `ErrorOr<T>` return types for file/directory-based operations should make the system more stable, once we start utilizing them fully.

I only noticed one main thing in `Core::Filesystem` itself that I would personally change, but it's mostly nitpicking, you don't have to:
- I feel like we could move the `TempFile` class into its own `.cpp` file, like it was before, maybe `LibCore/Filesystem/TempFile.cpp`, it feels a bit awkward with it hanging around with the "helper functions" in `Filesystem.cpp`.

Some points for the future which I would like to throw out there. Please note that none of these should be done by you specifically (unless you feel like you want to), or are related to this PR - they just are related to it and came across my mind whilst reviewing:
- `LexicalPath` is still quite dependent on `DeprecatedString`, and is also a bit bloated. We could honestly do a `DeprecatedFile` approach and make a new `Path` class, which is more mod,ern and slowly phase out `LexicalPath`.
- A new `Path` class could possibly be a gateway to these helpers, without having to throw around `String`s in functions and classes, yes, the `Path` would basically be a glorified String. It would however allow for doing `copy_file` or `remove` directly on the `Path` instead of having to call ``Core::Filesystem::remove(m_path, RecusrionMode::Disallowed);`` - which is still fine, but it's a bit verbose, especially if you're already using `Path` (or `LexicalPath`) somewhere.

Remember, those points above are just my opinion, and we should probably discuss what to do with `LexicalPath` more somewhere else.

Either way, this PR is a nice step forward and I can't wait to start using it in ImageViewer in particular, because right now it really needs some `String`, `Core::File`, and `Core::Filesystem` love!

**Review by kleinesfilmroellchen — CHANGES_REQUESTED** [review]

File: Userland/Applications/PartitionEditor/PartitionModel.h [hunk 22]
```diff
@@ -22,7 +22,7 @@ class PartitionModel final : public GUI::Model {
         __Count,
     };
 
-    static NonnullRefPtr<PartitionModel> create() { return adopt_ref(*new PartitionModel()); }
+    static NonnullRefPtr<PartitionModel> create();
```
Comment (linusg) [inline Userland/Applications/PartitionEditor/PartitionModel.h:25]: This seems completely unrelated to the commit

Comment (cammo1123) [reply Userland/Applications/PartitionEditor/PartitionModel.h:25]: Definitely seems that way at a first glance, but see my previous comment here. If this needs its own commit or something let me know!

>All were very simple changes to make. Done! Had to >move `PartitionModel::create()` out of line to get it to >stop complaining about the forward declaration of >`DeprecatedFile`
>
>~~The only issue is the `DeprecatedFile` header needs >to be declared in `PartitionModel.h` otherwise, it fails >to compile. (Complains about something to do with >ForwardDeclaring and unref-ing `DeprecatedFile`)~~

Comment (linusg) [reply Userland/Applications/PartitionEditor/PartitionModel.h:25]: ah, fair enough :)

**Comment by linusg** [comment]
> I don't think the minimal amount of code within TempFile warrants a change to the whole LibCore layout. (As in, LibCore currently doesn't have any subdirectories)

I fully agree with @caoimhebyrne, go create that subdirectory :)
Slight preference for something like `LibCore/FileSystem/Utils.{cpp,h}` over `LibCore/FileSystem.{cpp,h}` and `LibCore/FileSystem/*`. Things like `File` and `Directory` can be moved there as well, later on.

**Review by kleinesfilmroellchen — APPROVED** [review]

File: Userland/Libraries/LibFileSystem/TempFile.h [hunk 23]
```diff
@@ -20,7 +20,7 @@ class TempFile {
 
     ~TempFile();
 
-    String const& path() const { return m_path; }
+    StringView path() const { return m_path; }
```
[the remark this answers is not in the document]
Comment (timschumi) [reply Userland/Libraries/LibFileSystem/TempFile.h:23]: Doesn't that work either way? Like, where does a reference to a String get kept alive where a StringView wouldn't?

[the remark this answers is not in the document]
Comment (linusg) [reply Userland/Libraries/LibFileSystem/TempFile.h]: by 'keep alive' I mean increasing the refcount of the underlying `String` if we happen to copy it, not the actual C++ reference. If you return a string view the API turns into a footgun, if you return a String it is safe by default, storing a view into the returned string becomes the caller's choice.
